src/System.Private.CoreLib/src/System/Threading/WaitSubsystem.ThreadWaitInfo.Unix.cs

```diff
@@ -118,18 +118,40 @@ public WaitableObject[] GetWaitedObjectArray(int requiredCapacity)
                 Debug.Assert(_thread == Thread.CurrentThread);
                 Debug.Assert(_waitedCount == 0);
 
-                _waitedObjects.VerifyElementsAreDefault();
-                _waitedObjects.EnsureCapacity(requiredCapacity);
-                return _waitedObjects.Items;
+#if DEBUG
+                for (int i = 0; i < _waitedObjects.Length; ++i)
+                {
+                    Debug.Assert(_waitedObjects[i] == null);
+                }
+#endif
+
+                int currentLength = _waitedObjects.Length;
+                if (currentLength < requiredCapacity)
+                    _waitedObjects = new WaitableObject[Math.Max(requiredCapacity,
+                        Math.Min(WaitHandle.MaxWaitHandles, 4 + 2 * currentLength))];
+
+                return _waitedObjects;
             }
 
             private WaitedListNode[] GetWaitedListNodeArray(int requiredCapacity)
             {
                 Debug.Assert(_thread == Thread.CurrentThread);
                 Debug.Assert(_waitedCount == 0);
 
-                _waitedListNodes.EnsureCapacity(requiredCapacity, i => new WaitedListNode(this, i));
```

Comment:
This is where the delegate got allocated every time.

Reply:
Missed that one, thanks!